### PR TITLE
KAFKA-8807: Fix for potential test flakiness

### PR DESCRIPTION
**Pull request by bbejeck**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
ping @guozhangwang, @ableegoldman, @cadonna for review

**Comment by ableegoldman** [comment]
Thanks Bill! Build failed with environmental issues, retest this please.

Just so I understand, why might the order of records differ? (I assume the failure was due to comparing order and not just contents?)

**Comment by bbejeck** [comment]
>Just so I understand, why might the order of records differ? (I assume the failure was due to comparing order and not just contents?)

The test failure involved duplicate records.  I ran this locally 100 times and could not reproduce a failure.  Additionally, there's only been one reported failure of this test.  So I'm not convinced the failure exposed a real issue

EDIT: Having said that, I'm not sure how often users are using the `close` method, so I'll make a pass to make sure it's not a real issue.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java [hunk 1]
```diff
@@ -74,7 +75,7 @@
     private KafkaStreams kafkaStreams;
     private String globalStoreTopic;
     private String streamTopic;
-    private final List<Long> retrievedValuesList = new ArrayList<>();
+    private final Set<Long> retrievedValuesList = new HashSet<>();
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java:78]: nit: we should update the name too if we're changing this from a list

**Comment by ableegoldman** [comment]
Hm. I'm somewhat hesitant to ignore duplicate records, especially when the test itself is of a graceful shutdown. But maybe I'm biased because mysterious duplicate records helped me uncover a bug in the clients for 429 😛 
Of course if 100 runs don't turn up anything maybe there's no reason for concern. But I'm still not sure we should cover up the issue vs. just closing the ticket as `cannot reproduce`? WDYT (also @guozhangwang and @cadonna )?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java:78]: ack

**Comment by bbejeck** [comment]
>Of course if 100 runs don't turn up anything maybe there's no reason for concern. But I'm still not sure we should cover up the issue vs. just closing the ticket as cannot reproduce? WDYT (also @guozhangwang and @cadonna )?

Good point.  Given from what I said above
>Having said that, I'm not sure how often users are using the close method, so I'll make a pass to make sure it's not a real issue.

It's worth looking into it some and see what I can find.  If nothing then maybe close the ticket as you suggest.

**Comment by guozhangwang** [comment]
@bbejeck can we trigger a branch builder to ran this single test 100 times on jenkins to see if it can expose something?

**Comment by bbejeck** [comment]
It’s an integration test so I don’t think branch builder will work.

**Comment by guozhangwang** [comment]
You do not reuse the current scripted jobs, but can create your own job and the script just runs this single test :)

**Comment by bbejeck** [comment]
@guozhangwang @ableegoldman 

I've been thinking about this test some and I think I should refactor the test.  For context, the genesis of the `GlobalShutdownOrderTest` is from https://issues.apache.org/jira/browse/KAFKA-6844.

A race condition exists where shut down is called on a StreamThread then shut down is called on a GlobalStreamThread.   But the StreamThread can be delayed in shutting down,  and the GlobalStreamThread can shutdown first.

So the test should measure that the list is not empty and that `close`  is only executed once.
For the test failure with duplicate values, the only way there could be duplicate values is that Kafka Streams called `close`  twice.  

So I'm going to re-submit a PR with the changes mentioned above.  I'll also create a system test with the same topology and execute it 100x or so to see if there is an issue with close.

**Comment by guozhangwang** [comment]
> So I'm going to re-submit a PR with the changes mentioned above. I'll also create a system test with the same topology and execute it 100x or so to see if there is an issue with close.

+100, I agree with your refactoring motivation!

**Pull request closed without merging**